### DISCUSSION TITLE
FS-1056 - Allow overloads of custom keywords in computation expressions

### DISCUSSION THREAD

**Discussion by Nhowka** [fsharp/fslang-design#301] [Language and core library RFC discussions]
[RFC here](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1056-allow-custom-operation-overloads.md)

**Comment by smoothdeveloper** [comment]
Would this feature help with what the LINQ query builder can do?

https://github.com/dotnet/fsharp/issues/11446

Maybe the different functions to handle nullable value in LINQ expressions could be unified in later versions of FSharp.Core?

**Comment by Thorium** [comment]
As the builders are coming fashionable, we'd need more validation possibilities, and extending this would help.

Here is a code sample that is not working in F# 5.0:

```fsharp
type InitialState =
    { Name : string
      Username : string option
    }

type ValidState =
    { Name : string
      Username : string
      MoreData : string option
    }

type MyBuilder() =
    member __.Yield _ =
        { Name = ""
          Username = None
        } : InitialState

    // This fails for custom operator overloading now:
    [<CustomOperation "name">]
    member __.Name(state:InitialState, name) = { state with Name = name }
    member this.Name(state:ValidState, name) = { state with Name = name }

    [<CustomOperation "username">]
    member __.Username(state:InitialState, username) = 
        { Name = state.Name
          Username = username
          MoreData = None
        } : ValidState 

let my = type MyBuilder()
let process (x:ValidState) = 
    //...
    ()

my { name "hello"; username "hello" } |> process // Valid
my { username "hello"; name "hello" } |> process // Valid
my { name "hello" } |> process // Intentionally invalid, compile time error
```

**Reply by Nhowka** [reply]
Sorry, I didn't catch which part is not working. Is it about the compile time error? In that case, the builder is creating an `InitialState` while `process` is expecting a `ValidState`. I had to change the `Username` member to having `Username = Some username`, but then it worked fine for the first two cases.

**Reply by Thorium** [reply]
That's exactly what I'd want.
What I cannot do is the Name field for both valid and invalid states.

**Reply by Nhowka** [reply]
What error are you getting? I was testing it on the fable repl and I feel that it's working fine. [Check it here](https://fable.io/repl/#?gist=12e3f83a463ae99f150e7e859e121358)

**Reply by Thorium** [reply]
The context here:
Currently [Farmer](https://compositionalit.github.io/farmer/) gathers the state via computation expression, and then when executed, it generates an ARM-template.

```fsharp
vm {
   name "abc"
   username "myuser"
   system_identity
}
```

The thing I don't like, is that it validates the state at the execution (runtime), meanwhile your typical F# programs would make the illegal state un-representable:

[`failwith $"You must specify a username for virtual machine {this.Name.Value}"`](https://github.com/CompositionalIT/farmer/blob/bbc7600ad4039584506b7f4c0152ca59b83213e1/src/Farmer/Builders/Builders.Vm.fs#L72)

But the Farmer is not taking view to the command order (and for backward-compatibility it should stay as that).
So what we want: InitialState: no-username, ValidState: has username.
Very [simple change](https://github.com/Thorium/farmer/commit/96df40306f32d9619a90a08544d1ca3b9e986214) for the that,
here is the full repo: https://github.com/Thorium/farmer/tree/operator-overload

And what happens, when I try to build:
`error FS3087: The custom operation 'name' refers to a method which is overloaded. The implementations of custom operations may not be overloaded.`

![farmer](https://user-images.githubusercontent.com/229355/125269569-bc161100-e300-11eb-9cf3-e22464110e1f.png)

**Reply by Thorium** [reply]
Ok, found it!
All the overloads have to be bundled together, you cannot do this:
```
    [<CustomOperation "name">]
    member this.Name(state:ValidState, name) = { state with Name = name }

    [<CustomOperation "username">]
    member __.Username(state:InitialState, username) = 
        { Name = state.Name
          Username = Some username
          MoreData = None
        } : ValidState 

    [<CustomOperation "name">]
    member __.Name(state:InitialState, name) = { state with Name = name }
```

Ok then...

**Reply by Thorium** [reply]
Another place where this seems to be a problem is that if the custom operator is coming via an interface:
![identity](https://user-images.githubusercontent.com/229355/125276959-c3d9b380-e308-11eb-9c18-2876f6fedea8.png)

**Reply by Nhowka** [reply]
When testing it here it seems like it's in a weird middle state. It works when you add an overload but it breaks with the `The implementations of custom operations may not be overloaded.` error when you add the `CustomOperation` attribute again on that second method.

It worked fine and the error went away when I changed the lang version to preview, but probably the code that is working shouldn’t work without that flag. My hope is that that kind of weird behavior disappears when the feature comes out of preview.
